File: Arrays_and_Hashing/Custom_Comparators_and_Coordinate_Compression/Silver/P1_2018-Out_of_Sorts/P1_2018-Out_of_Sorts.py
import typing, functools

class Entry(typing.NamedTuple):
	value: int
	index: int

with open("sort.in", "r") as fin:
	# Read in first line: n = length of array
	n = int(fin.readline())
	# For the next n lines, record the value & index of each entry
	entries = [Entry(int(fin.readline()), i) for i in range(n)]

# Sort the array in ascending order, breaking ties by initial index
comparator = (
	lambda a, b: a.index - b.index 
	if a.value == b.value 
	else a.value - b.value
)
entries.sort(key=functools.cmp_to_key(comparator))

# Using initial indeces count the # of bubbles needed to sort the array
print(
	max(entries[i].index - i + 1 for i in range(n)), 
	file=open("sort.out", "w")
)

File: Arrays_and_Hashing/Custom_Comparators_and_Coordinate_Compression/Silver/P1_2018-Out_of_Sorts/P1_2018-Out_of_Sorts.cpp
#include <bits/stdc++.h>

#define N_MAX 100001

using namespace std;

struct Entry {
	int value, index;
};

int main() {
	freopen("sort.in", "r", stdin);
	// Read in first line: n = length of array
	int n;
	cin >> n;
	// For the next n lines, record the value & index of each entry
	Entry entries[N_MAX];
	for (int idx = 0; idx < n; idx++) {
		cin >> entries[idx].value;
		entries[idx].index = idx;
	}

	// Sort the array in ascending order, breaking ties by initial index
	sort(entries, entries + n, 
		[](const Entry &a, const Entry &b) {
			if (a.value == b.value) 
				return a.index < b.index;
			return a.value < b.value;
		}
	);

	// Using initial indeces count the # of bubbles needed to sort the array
	int mooCount = 1;
	for (int i = 0; i < n; i++) 
		mooCount = max(mooCount, entries[i].index - i + 1);
		// add 1 to account for the final iteration of the algorithm
	// Write the number of "moos"/bubbles needed to sort the array
	freopen("sort.out", "w", stdout);
	cout << mooCount << endl;
}
